File: Grammars/Assignment/MarkovChainGenerator.vb
Namespace Assignment
    Public Class MarkovChainGenerator

        Private START_LABEL As String = "Start"

        Private _grammar As AssignmentGrammar
        Private _samples As IEnumerable(Of String)
        Private _syllableLength As Int32
        Private _weightLimit As Int32 = 0

        Public Sub New(ByVal samples As IEnumerable(Of String),
                       ByVal syllableLength As Int32,
                       ByVal weightLimit As Int32)
            _samples = samples
            _syllableLength = syllableLength
            _weightLimit = weightLimit
        End Sub

        Public Function BuildGrammar() As AssignmentGrammar
            Dim list As New List(Of KeyValuePair(Of String, String))
            _grammar = New AssignmentGrammar

            For Each sample As String In _samples
                Dim i As Int32 = 0
                Dim diff As Int32 = sample.Length Mod _syllableLength
                Dim previous As String = START_LABEL
                Dim current As String

                If diff <> 0 Then
                    sample &= Space(diff)
                End If
                Do While i < sample.Length
                    current = sample.Substring(i, _syllableLength).Trim
                    list.Add(New KeyValuePair(Of String, String)(previous, current))
                    If i = sample.Length - _syllableLength Then
                        list.Add(New KeyValuePair(Of String, String)(current, String.Empty))
                    End If
                    previous = current.Trim
                    i += _syllableLength
                Loop
            Next

            For Each pair As KeyValuePair(Of String, String) In list
                Dim rule As LineItem = _grammar.Rules.FirstOrDefault(Function(r As LineItem)
                                                                         If r.Name = If(String.IsNullOrWhiteSpace(pair.Key), "Start", pair.Key) AndAlso
                                                                            If(String.IsNullOrEmpty(r.Next), r.Expression = String.Format("[{0}]", pair.Value), r.Next = pair.Value) Then Return True
                                                                         Return False
                                                                     End Function)
                If rule IsNot Nothing Then
                    If _weightLimit = 0 OrElse CInt(rule.Weight) < _weightLimit Then rule.Weight += 1
                Else
                    rule = New LineItem(pair.Key,
                                        If(pair.Key = START_LABEL, String.Empty, pair.Value),
                                        If(pair.Key = START_LABEL, String.Format("[{0}]", pair.Value), pair.Key))
                    rule.Weight = 1
                    _grammar.Rules.Add(rule)
                End If
            Next

            Return _grammar
        End Function

    End Class
End Namespace

File: Grammars/Assignment/MarkovChainGenerator.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualBasic;

namespace Grammars.Assignment
{
    public class MarkovChainGenerator
    {
        private readonly IEnumerable<string> _samples;
        private readonly int _syllableLength;
        private readonly int _weightLimit;

        private readonly string START_LABEL = "Start";
        private AssignmentGrammar _grammar;

        public MarkovChainGenerator(IEnumerable<string> samples, int syllableLength, int weightLimit)
        {
            _samples = samples;
            _syllableLength = syllableLength;
            _weightLimit = weightLimit;
        }

        public AssignmentGrammar BuildGrammar()
        {
            var list = new List<KeyValuePair<string, string>>();
            _grammar = new AssignmentGrammar();

            foreach (var s in _samples)
            {
                var sample = s;

                var i = 0;
                var diff = sample.Length%_syllableLength;
                var previous = START_LABEL;

                if (diff != 0)
                {
                    sample += Strings.Space(diff);
                }
                while (i < sample.Length)
                {
                    var current = sample.Substring(i, _syllableLength).Trim();
                    list.Add(new KeyValuePair<string, string>(previous, current));
                    if (i == sample.Length - _syllableLength)
                    {
                        list.Add(new KeyValuePair<string, string>(current, string.Empty));
                    }
                    previous = current.Trim();
                    i += _syllableLength;
                }
            }

            foreach (var pair in list)
            {
                var rule = _grammar.Rules.FirstOrDefault(
                    r =>
                    {
                        if (r.Name == ((string.IsNullOrWhiteSpace(pair.Key)) ? "Start" : pair.Key) &&
                            ((string.IsNullOrEmpty(r.Next))
                                ? (r.Expression == string.Format("[{0}]", pair.Value))
                                : r.Next == pair.Value))
                        {
                            return true;
                        }
                        return false;
                    });
                if (rule != null)
                {
                    if (_weightLimit == 0 || rule.Weight < _weightLimit)
                    {
                        rule.Weight++;
                    }
                }
                else
                {
                    rule = new LineItem(
                        pair.Key,
                        pair.Key == START_LABEL ? string.Empty : pair.Value,
                        pair.Key == START_LABEL ? string.Format("[{0}]", pair.Value) : pair.Key);
                    rule.Weight = 1;
                    _grammar.Rules.Add(rule);
                }
            }

            return _grammar;
        }
    }
}